File: v2/src/bun/services/ready-url.ts
/**
 * Ready-URL detection from process log lines.
 *
 * Patterns are stored in SQLite and managed via Settings → Log URL patterns.
 * Each pattern runs against a single ANSI-stripped log line (no trailing newline).
 *
 * Capture groups (use named groups):
 * - `url`  — full URL (preferred), e.g. http://localhost:5173/
 * - `port` — port only; becomes http://localhost:{port}
 *
 * If both are present, `url` wins.
 */

import { readyUrlPatternsRepo } from "../db/ready-url-patterns"

export type ReadyUrlPattern = {
  id: string
  label: string
  pattern: RegExp
}

export type ReadyUrlMatch = {
  url: string
  patternId: string
  label: string
}

let cached: ReadyUrlPattern[] | null = null

export function invalidateReadyUrlPatternsCache() {
  cached = null
}

export function validateReadyUrlPattern(
  source: string,
  flags = "i"
): { ok: true; pattern: RegExp } | { ok: false; error: string } {
  let re: RegExp
  try {
    re = new RegExp(source, flags)
  } catch (err) {
    return {
      ok: false,
      error: err instanceof Error ? err.message : "Invalid regular expression",
    }
  }

  if (!/\(\?<url>/.test(source) && !/\(\?<port>/.test(source)) {
    return {
      ok: false,
      error: "Pattern must include a named group `url` and/or `port`",
    }
  }

  return { ok: true, pattern: re }
}

function loadPatterns(): ReadyUrlPattern[] {
  readyUrlPatternsRepo.ensureSeeded()
  const rows = readyUrlPatternsRepo.list()
  const compiled: ReadyUrlPattern[] = []
  for (const row of rows) {
    try {
      compiled.push({
        id: String(row.id),
        label: row.label,
        pattern: new RegExp(row.pattern, row.flags),
      })
    } catch {
      // Skip broken rows so one bad pattern doesn't break matching.
    }
  }
  return compiled
}

function getPatterns(): ReadyUrlPattern[] {
  if (!cached) cached = loadPatterns()
  return cached
}

/** Trim trailing junk that often sticks to URLs in logs. */
function cleanUrl(raw: string): string {
  return raw.replace(/[)\].,;:'">]+$/g, "")
}

function fromPort(port: string): string | null {
  const n = Number(port)
  if (!Number.isInteger(n) || n < 1 || n > 65535) return null
  return `http://localhost:${n}`
}

/**
 * Try every configured pattern against one log line.
 * Returns the first match, or null.
 */
export function matchReadyUrl(line: string): ReadyUrlMatch | null {
  const text = line.trim()
  if (!text) return null

  for (const { id, label, pattern } of getPatterns()) {
    pattern.lastIndex = 0
    const m = pattern.exec(text)
    if (!m?.groups) continue

    const urlGroup = m.groups.url
    if (urlGroup) {
      const url = cleanUrl(urlGroup)
      if (/^https?:\/\//i.test(url)) {
        return { url, patternId: id, label }
      }
      continue
    }

    const portGroup = m.groups.port
    if (portGroup) {
      const url = fromPort(portGroup)
      if (url) return { url, patternId: id, label }
    }
  }

  return null
}

File: wails_backend/internal/services/readyurl.go
package services

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"wails_backend/internal/db"
)

type ReadyURLMatch struct {
	URL       string
	PatternID string
	Label     string
}

type compiledPattern struct {
	id      string
	label   string
	pattern *regexp.Regexp
}

var (
	readyMu    sync.Mutex
	readyCache []compiledPattern
)

func InvalidateReadyURLPatternsCache() {
	readyMu.Lock()
	readyCache = nil
	readyMu.Unlock()
}

func jsFlagsToPrefix(flags string) string {
	var chars []byte
	seen := map[rune]bool{}
	for _, c := range flags {
		if seen[c] {
			continue
		}
		switch c {
		case 'i', 'm', 's':
			seen[c] = true
			chars = append(chars, byte(c))
		}
	}
	if len(chars) == 0 {
		return ""
	}
	return "(?" + string(chars) + ")"
}

func jsNamedGroups(src string) string {
	return regexp.MustCompile(`\(\?<(\w+)>`).ReplaceAllString(src, "(?P<$1>")
}

func CompileJSRegexp(source, flags string) (*regexp.Regexp, error) {
	return regexp.Compile(jsFlagsToPrefix(flags) + jsNamedGroups(source))
}

func ValidateReadyURLPattern(source, flags string) error {
	if _, err := CompileJSRegexp(source, flags); err != nil {
		return fmt.Errorf("Invalid regular expression: %s", err.Error())
	}
	if !strings.Contains(source, "(?<url>") && !strings.Contains(source, "(?<port>") {
		return fmt.Errorf("Pattern must include a named group `url` and/or `port`")
	}
	return nil
}

func loadPatterns(ctx context.Context, d *db.DB) []compiledPattern {
	_ = d.EnsureReadyURLPatternsSeeded(ctx)
	rows, err := d.ListReadyUrlPatterns(ctx)
	if err != nil {
		return nil
	}
	var compiled []compiledPattern
	for _, row := range rows {
		re, err := CompileJSRegexp(row.Pattern, row.Flags)
		if err != nil {
			continue
		}
		compiled = append(compiled, compiledPattern{
			id:      strconv.FormatInt(row.ID, 10),
			label:   row.Label,
			pattern: re,
		})
	}
	return compiled
}

func getPatterns(ctx context.Context, d *db.DB) []compiledPattern {
	readyMu.Lock()
	defer readyMu.Unlock()
	if readyCache == nil {
		readyCache = loadPatterns(ctx, d)
	}
	return readyCache
}

func cleanURL(raw string) string {
	return strings.TrimRight(raw, ")].,;:'\">")
}

func fromPort(port string) string {
	n, err := strconv.Atoi(port)
	if err != nil || n < 1 || n > 65535 {
		return ""
	}
	return fmt.Sprintf("http://localhost:%d", n)
}

func MatchReadyURL(ctx context.Context, d *db.DB, line string) *ReadyURLMatch {
	text := strings.TrimSpace(line)
	if text == "" {
		return nil
	}
	for _, p := range getPatterns(ctx, d) {
		m := p.pattern.FindStringSubmatch(text)
		if m == nil {
			continue
		}
		names := p.pattern.SubexpNames()
		groups := map[string]string{}
		for i, name := range names {
			if i == 0 || name == "" || i >= len(m) {
				continue
			}
			groups[name] = m[i]
		}
		if urlGroup := groups["url"]; urlGroup != "" {
			url := cleanURL(urlGroup)
			if strings.HasPrefix(strings.ToLower(url), "http://") || strings.HasPrefix(strings.ToLower(url), "https://") {
				return &ReadyURLMatch{URL: url, PatternID: p.id, Label: p.label}
			}
			continue
		}
		if portGroup := groups["port"]; portGroup != "" {
			if url := fromPort(portGroup); url != "" {
				return &ReadyURLMatch{URL: url, PatternID: p.id, Label: p.label}
			}
		}
	}
	return nil
}
